sotodlib/toast/workflows/so_map.py

```diff
@@ -0,0 +1,253 @@
+
+    wrk.mapmaker_ml(job, otherargs, runargs, data)
+
+    wrk.filter_ground(job, otherargs, runargs, data)
+    wrk.filter_poly1d(job, otherargs, runargs, data)
+    wrk.filter_poly2d(job, otherargs, runargs, data)
+    wrk.filter_common_mode(job, otherargs, runargs, data)
+
+    wrk.mapmaker(job, otherargs, runargs, data)
+    wrk.mapmaker_filterbin(job, otherargs, runargs, data)
+    wrk.mapmaker_madam(job, otherargs, runargs, data)
+
+    wrk.filtered_statistics(job, otherargs, runargs, data)
+
+    mem = toast.utils.memreport(
+        msg="(whole node)", comm=data.comm.comm_world, silent=True
+    )
+    log.info_rank(f"After reducing data:  {mem}", data.comm.comm_world)
+
+
+def load_data(job, otherargs, runargs, data):
+    log = toast.utils.Logger.get()
+    job_ops = job.operators
+
+    # Timer for reporting the progress
+    timer = toast.timing.Timer()
+    timer.start()
+
+    job_ops.mem_count.prefix = "Before Data Load"
+    job_ops.mem_count.apply(data)
+
+    # Load data from all formats
+    wrk.load_data_hdf5(job, otherargs, runargs, data)
+    wrk.load_data_books(job, otherargs, runargs, data)
+    # wrk.load_data_context(job, otherargs, runargs, data)
+
+    if len(data.obs) == 0:
+        raise RuntimeError("No input data specified!")
+
+    job_ops.mem_count.prefix = "After Data Load"
+    job_ops.mem_count.apply(data)
+
+
+def main():
+    env = toast.utils.Environment.get()
+    log = toast.utils.Logger.get()
+    gt = toast.timing.GlobalTimers.get()
+    gt.start("toast_so_map (total)")
+    timer = toast.timing.Timer()
+    timer.start()
+
+    # Get optional MPI parameters
+    comm, procs, rank = toast.get_world()
+
+    if "OMP_NUM_THREADS" in os.environ:
+        nthread = os.environ["OMP_NUM_THREADS"]
+    else:
+        nthread = 1
+    log.info_rank(
+        f"Executing workflow with {procs} MPI tasks, each with "
+        f"{nthread} OpenMP threads at {datetime.datetime.now()}",
+        comm,
+    )
+
+    mem = toast.utils.memreport(msg="(whole node)", comm=comm, silent=True)
+    log.info_rank(f"Start of the workflow:  {mem}", comm)
+
+    parser = argparse.ArgumentParser(description="SO mapmaking pipeline")
+
+    parser.add_argument(
+        "--out_dir",
+        required=False,
+        type=str,
+        default="toast_out",
+        help="The output directory",
+    )
+    parser.add_argument(
+        "--obsmaps",
+        required=False,
+        default=False,
+        action="store_true",
+        help="Map each observation separately.",
+    )
+
+    # The operators and templates we want to configure from the command line
+    # or a parameter file.
+
+    operators = list()
+    templates = list()
+
+    wrk.setup_load_data_hdf5(operators)
+    wrk.setup_load_data_books(operators)
+    wrk.setup_load_data_context(operators)
+
+    wrk.setup_pointing(operators)
+    wrk.setup_simple_noise_models(operators)
+    wrk.setup_flag_noise_outliers(operators)
+
+    wrk.setup_filter_hwpss(operators)
```

Comment:
This is obviously infinitely better than the old, super long workflow script. I still wonder, if we could set up all the canonical data reduction operators with a single `wrk.setup_data_reduction()`? I'm concerned that when we develop a new data reduction operator, we will have to edit a large number of workflows that all do data reduction.

Reply:
In the model where we have a few giant scripts doing everything, there are a couple places to edit.  In the model where we have many smaller scripts doing specific things, my thought was that these would have a very specific set of simulation or reduction steps and we would not necessarily add new operators to everything (otherwise we would end up with the giant script model).  Taking this to the logical extreme, one could write a single workflow that just took a bunch of config files and executed things in a particular order and then stop writing scripts and just edit toml/json files.  I was aiming for some middle ground, where we can describe (different) types of high-level data flow in each script.

Reply:
Fair enough. So in this model we anticipate a relatively small number of workflows that could require synchronous editing when the canonical set of data reduction operators is augmented.

Reply:
I think we will have both.  Some people prefer a giant script that does everything with pieces enabled / disabled by config files.  Others prefer a script with only the steps for a specific use case (e.g. SAT transfer function with loading from a context and calling wrapped preprocessing, or simulated observing with 3 different mapmakers running on the same data, etc).  So in the do-everything-script case there are just a couple places to add new things.  In the other case there might be one or two specific scripts that need updating, but not all of them.

Perhaps later in the project, when there is a more stable reduction sequence, we could make another helper function that wraps even more things.  Or if we find many workflows have the same sequence of things, we could extract that out.  I am just advocating not doing that until it becomes clear what those blocks are.